### PR TITLE
Use a LinkedList to improve the performances of ChangeNotifier

### PR DESCRIPTION
**Pull request by rrousselGit**
## Description

This PR improves the performances of ChangeNotifier by using a `LinkedList` to store the listeners instead of an `ObserverList`.
Such change avoids having to do a `contains` call to know if a listener is still present or not.

I made a benchmark to compare both versions: https://gist.github.com/rrousselGit/5a047bd4ec36515a4cfcc6bd275f05f5

This benchmark measures the time `notifyListeners` takes to complete based on the number of listeners the notifier has.
The results I obtained can be seen in this table (performed before NNBD migration):

run 1:

|        | 0 listeners            | 1                     | 2                      | 3                     | 4                     | 5                      |
|--------|------------------------|-----------------------|------------------------|-----------------------|-----------------------|------------------------|
| Before |91.1281665755422 us. | 381.1135670731707 us. |828.2728778467908 us.  | 899.8349077822762 us.         |  1535.7022256331543 us. |1767.6510600706713 us. |
| After  | 8.112791806104857 us.  |183.02415812591508 us. | 235.99882005899704 us. | 291.88019845323214 us. | 386.16795366795367 us. |439.40685413005275 us.  |

run 2:

|        | 0 listeners           | 1                      | 2                      | 3                      | 4                     | 5                     |
|--------|-----------------------|------------------------|------------------------|------------------------|-----------------------|-----------------------|
| Before |91.34866407855675 us. |  381.59557420831743 us. |  825.4416013206768 us. | 1066.5165245202559 us. |  1690.8469991546915 us. | 1578.2799684542588 us. |
| After  |  8.114009606958554 us. | 179.82064191315294 us. |232.83119906868453 us. | 287.4411553384107 us. |  380.2490494296578 us. | 434.25814155449416 us. |

It seems like even for a small number of listeners, the changes proposed are still faster. And the difference grows exponentially, as `notifyListeners` is now O(N) instead of O(N^2)

I'm not a benchmark guru though, so the benchmark I linked above may be flawed. Feel free to correct me.

## Related Issues

#59611

## Tests

NaN, this PR does not contain any behavior change.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 1]
```diff
@@ -93,18 +94,25 @@ abstract class ValueListenable<T> extends Listenable {
   T get value;
 }
 
+class _ListenerEntry extends LinkedListEntry<_ListenerEntry> {
+  _ListenerEntry(this.listener);
+  final void Function() listener;
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/change_notifier.dart:99]: use a typedef here.

Comment (goderbauer) [reply packages/flutter/lib/src/foundation/change_notifier.dart:99]: Probably just `VoidCallback`?

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 2]
```diff
@@ -93,18 +94,25 @@ abstract class ValueListenable<T> extends Listenable {
   T get value;
 }
 
+class _ListenerEntry extends LinkedListEntry<_ListenerEntry> {
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/change_notifier.dart:97]: This probably deserves a comment explaining why we have this wrapper object instead of just using the `listener` directly.

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 3]
```diff
@@ -171,7 +181,13 @@ class ChangeNotifier implements Listenable {
...
   void removeListener(VoidCallback listener) {
     assert(_debugAssertNotDisposed());
-    _listeners!.remove(listener);
+    _listeners
+        !.cast<_ListenerEntry?>()
+        .firstWhere(
+          (_ListenerEntry? element) => element!.listener == listener,
+          orElse: () => null,
+        )
+        ?.unlink();
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/change_notifier.dart:190]: Wondering if this would be even faster with a regular for loop. (The code would definitely be prettier without the casting).

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 4]
```diff
@@ -205,12 +221,13 @@ class ChangeNotifier implements Listenable {
   @visibleForTesting
   void notifyListeners() {
     assert(_debugAssertNotDisposed());
-    if (_listeners != null) {
-      final List<VoidCallback> localListeners = List<VoidCallback>.from(_listeners!);
-      for (final VoidCallback listener in localListeners) {
+    _firstNewlyAddedEntry = null;
+    if (_listeners != null && _listeners!.isNotEmpty) {
+      for (_ListenerEntry? entry = _listeners!.first;
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/change_notifier.dart:226]: Why's entry nullable?

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 5]
```diff
@@ -205,12 +221,13 @@ class ChangeNotifier implements Listenable {
...
+            entry != null && entry != _firstNewlyAddedEntry;
+            entry = entry.next) {
         try {
-          if (_listeners!.contains(listener))
-            listener();
+          entry.listener();
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/change_notifier.dart:230]: What if `entry.listener()` calls `removeListener` for itself? Wouldn't that set `entry.next` to null and the iteration would end early?

Comment (goderbauer) [reply packages/flutter/lib/src/foundation/change_notifier.dart:230]: (If we don't have a test for this scenario, we should add one)

**Comment by goderbauer** [comment]
It would probably be best to submit the benchmark fist to https://github.com/flutter/flutter/tree/master/dev/benchmarks/microbenchmarks so we can collect a couple of numbers with the current implementation and then see how this PR changes the numbers when it gets submitted.

The benchmark should probably also track the performance of the other relevant operations (adding/removing listeners).

File: packages/flutter/lib/src/foundation/change_notifier.dart (hunk 2 above)
Comment (rrousselGit) [reply packages/flutter/lib/src/foundation/change_notifier.dart:97]: It is required by LinkedList. We cannot do `LinkedList<VoidCallback>`.

We could change `LinkedList<_ListenerEntry>` by `DoubleLinkedQueue<VoidCallback>` maybe?

File: packages/flutter/lib/src/foundation/change_notifier.dart (hunk 4 above)
Comment (rrousselGit) [reply packages/flutter/lib/src/foundation/change_notifier.dart:226]: `entry.next` may be `null` when reaching the end of the list. We need it to be nullable for the `entry = entry.next`

File: packages/flutter/lib/src/foundation/change_notifier.dart (hunk 2 above)
Comment (goderbauer) [reply packages/flutter/lib/src/foundation/change_notifier.dart:97]: Ah, I see. I didn't look at the interface for LinkedList. Never mind.

File: packages/flutter/lib/src/foundation/change_notifier.dart (hunk 4 above)
Comment (goderbauer) [reply packages/flutter/lib/src/foundation/change_notifier.dart:226]: Makes sense.

**Comment by escamoteur** [comment]
I'm currently working on a package that offers rx like operators for ValueListenables so to say the poor mans rx without streams.
https://github.com/escamoteur/listenable_pipe/tree/master
In this context I would appreciate any performance improvements here

File: packages/flutter/lib/src/foundation/change_notifier.dart (hunk 5 above)
Comment (rrousselGit) [reply packages/flutter/lib/src/foundation/change_notifier.dart:230]: Good catch

I should add tests for `addListener(() { notifier.notifyListeners() })` too

**Comment by rrousselGit** [comment]
I've fixed all comments

What's remaining is:
- make a benchmark on https://github.com/flutter/flutter/tree/master/dev/benchmarks/microbenchmarks
- support recursive notifyListeners calls 

I'll come back to this PR in a bit later this week

**Comment by escamoteur** [comment]
It would be great if @munificent could have a look at the benchmark in respect of cache misses as we talked about on Twitter. As @letsar pointed out it may be that the whole benchmark data might fit in the cache so that we don't see any cache miss effects.

**Comment by goderbauer** [comment]
@escamoteur Can you link to the twitter conversation for context?

**Comment by rrousselGit** [comment]
https://twitter.com/lets4r/status/1287670675801604096?s=19

**Comment by dkwingsmt** [comment]
We probably want to make the same change to `actions.dart` where the exact listener logic is replicated.

(We might want to do the same to `InkWell` too, where `ObserverList` is used because we want to maintain a expectedly short list; however your benchmark shows that `LinkedList` beats `ObserverList` even in small number cases.)

**Comment by rrousselGit** [comment]
I've changed a bit the algorithm, to still rely on cloning the list of listeners
Using LinkedList still removes the `contains` call, so we are still winning in even for 1-2 listeners

The difference at 0 listeners is explained by the early return if there is no listener (as the current implementation do a list clone for no reason)

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 6]
```diff
@@ -93,18 +94,22 @@ abstract class ValueListenable<T> extends Listenable {
...
+  _ListenerEntry(this.listener);
+  final VoidCallback listener;
+}
+
 /// A class that can be extended or mixed in that provides a change notification
 /// API using [VoidCallback] for notifications.
 ///
-/// [ChangeNotifier] is optimized for small numbers (one or two) of listeners.
-/// It is O(N) for adding and removing listeners and O(N²) for dispatching
+/// It is O(1) for adding listeners and O(N) removing listeners and dispatching
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/change_notifier.dart:105]: ... and O(N) _for_ removing ...

File: packages/flutter/test/foundation/change_notifier_test.dart [hunk 7]
```diff
@@ -139,6 +139,67 @@ void main() {
...
+    source.addListener(listener1);
+
+    source.notify();
+
+    expect(log, <String>['listener1']);
+  });
+
+  test(
+      'If a listener in the middle of the list of listeners removes itself'
+      ', notifyListeners still notifies all listeners', () {
```
Comment (goderbauer) [inline packages/flutter/test/foundation/change_notifier_test.dart:164]: nit: the comma should be part of the previous line.

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 8]
```diff
@@ -205,27 +215,26 @@ class ChangeNotifier implements Listenable {
...
-            informationCollector: () sync* {
-              yield DiagnosticsProperty<ChangeNotifier>(
-                'The $runtimeType sending notification was',
-                this,
-                style: DiagnosticsTreeStyle.errorProperty,
-              );
-            },
-          ));
-        }
+    final List<_ListenerEntry> listeners = _listeners!.toList(growable: false);
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/change_notifier.dart:218]: nit: let's call this localListeners again to avoid confusion.

**Review by goderbauer — COMMENTED** [review]
This looks pretty interesting. 

Can you rebase this ontop the latest master to resolve the test failures?

Are the benchmark numbers in the description upto date with the latest implementation?

**Comment by goderbauer** [comment]
> The difference at 0 listeners is explained by the early return if there is no listener (as the current implementation do a list clone for no reason)

Which early return do you mean?

**Comment by rrousselGit** [comment]
> > The difference at 0 listeners is explained by the early return if there is no listener (as the current implementation do a list clone for no reason)
> 
> Which early return do you mean?

Inside notifyListeners

**Comment by rrousselGit** [comment]
Will do.
The benchmark numbers are indeed up to date with the latest implementation

**Comment by goderbauer** [comment]
> Inside notifyListeners

I don't see an early return in there?

**Comment by rrousselGit** [comment]
Oh I forgot to push it!

It's basically a:

```dart
notifyListeners() {
  if (listeners.isEmpty)
    return;
}
```

File: packages/flutter/test/foundation/change_notifier_test.dart [hunk 9]
```diff
@@ -139,6 +139,67 @@ void main() {
...
+    source.addListener(listener1);
+
+    source.notify();
+
+    expect(log, <String>['listener1']);
+  });
+
+  test(
+      'If a listener in the middle of the list of listeners removes itself,'
+      ' notifyListeners still notifies all listeners', () {
```
Comment (goderbauer) [inline packages/flutter/test/foundation/change_notifier_test.dart:164]: nit, the " " (space) should be moved to the previous line.

**Review by goderbauer — COMMENTED** [review]
The logic now looks sound to me.

We should submit the benchmark (https://github.com/flutter/flutter/pull/62715) first and let it collect a couple of data points before submitting this.

**Comment by goderbauer** [comment]
Not sure why that one check keeps failing on cirrus. You may have to rebase the PR to the latest master.

**Comment by dnfield** [comment]
Very cool! Approach seems sound to me. It'll have a slightly higher memory cost, but only slightly and worth it for the speed gain.

**Comment by rrousselGit** [comment]
Note that I still believe we should fuse `addListener` and `removeListener` as I mentioned in the issue (#59611) such that we have a usage similar to `Stream.listen` and its `StreamSubscription`.

This would fix the comment about `removeListener` having a surprising behavior when a listener is added multiple times:

>   /// If a listener had been added twice, and is removed once during an
  /// iteration (i.e. in response to a notification), it will still be called
  /// again. If, on the other hand, it is removed as many times as it was
  /// registered, then it will no longer be called. This odd behavior is the
  /// result of the [ChangeNotifier] not being able to determine which listener
  /// is being removed, since they are identical, and therefore conservatively
  /// still calling all the listeners when it knows that any are still
  /// registered.
  ///
  /// This surprising behavior can be unexpectedly observed when registering a
  /// listener on two separate objects which are both forwarding all
  /// registrations to a common upstream object.

**Comment by Hixie** [comment]
This PR looks great!

Would it make sense to deprecate ObserverList entirely if it is consistently slower than the LinkedList approach?

What platforms did we test this on? Might make sense to double-check the findings on iOS, x86, ARM, and Web to make sure it's better on all the platforms (could be a weird artifact of the platform it was tested on).

File: packages/flutter/test/foundation/change_notifier_test.dart [hunk 10]
```diff
@@ -139,6 +139,67 @@ void main() {
     log.clear();
   });
 
+  test('During notifyListeners, a listener was added and removed immediatly', () {
```
Comment (Hixie) [inline packages/flutter/test/foundation/change_notifier_test.dart:142]: nit - typo "immediately"

**Comment by rrousselGit** [comment]
I would recommend some of the team to run the benchmark themselves to double-check. As I mentioned, I'm not a benchmark expert and may have made a mistake when generating these numbers

For reference, the benchmark was added to the microbenchmark folder.

**Comment by Hixie** [comment]
It looks like nobody has the immediate bandwidth to set up benchmarking for this.

My recommendation would be to run some of our existing benchmarks on a couple of platforms (e.g. desktop, web, android, or whatever you have available), and see how those benchmarks fare. Maybe instrument some of the tests to report how many ChangeNotifiers they are using to find a benchmark that uses them more heavily than the others. (I'm not up to date on what benchmarks we have these days so I don't know which one to recommend off hand.)

**Comment by dnfield** [comment]
@Hixie is https://github.com/flutter/flutter/pull/62715 not satisfactory to benchmark this?

Those get run on Android and iOS (64 bit).

**Comment by Hixie** [comment]
I had understood @rrousselGit's comments above as saying that he didn't have confidence in his benchmark and wanted another. If we have a benchmark we have confidence in then great, let's just run that on a few platforms. :-)

**Comment by dnfield** [comment]
Between the specific microbenchmark and benchmarks we have for other tests on other platforms, I'd be confident in landing this as is.

**Review by goderbauer — APPROVED** [review]
LGTM

I am also pretty confident in landing this as is with the benchmark coverage we have.

**Comment by dnfield** [comment]
I landed this. We can keep an eye on the benchmarks for this commit in the devicelab. Assuming it goes well, we should look to apply this pattern to other places that could use it:

- RawKeyboard._handleKeyEvent (raw_keyboard.dart)
- AnimationLocalListenersMixen.notifyListeners (listener_helpers.dart)
- Action.notifyActionListeners (actions.dart)
- FocusManager._notifyHighlightModeListeners (focus_manager.dart).

Would be nice if we could somehow encapsulate the pattern for those.

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 11]
```diff
@@ -205,27 +215,29 @@ class ChangeNotifier implements Listenable {
...
-              );
-            },
-          ));
-        }
+    if (_listeners!.isEmpty)
+      return;
+
+    final List<_ListenerEntry> localListeners = List<_ListenerEntry>.from(_listeners!);
+
+    for (final _ListenerEntry entry in localListeners) {
```
Comment (dnfield) [inline packages/flutter/lib/src/foundation/change_notifier.dart:223]: Could we avoid this copy by getting the next entry before we invoke the current entry's listener?

Comment (dkwingsmt) [reply packages/flutter/lib/src/foundation/change_notifier.dart:223]: I wonder if any modification invalidates the list (theoretically). Iterable says modification during iteration is generally not allowed and may throw errors, although not necessarily for List. However can we do a lazy copy, i.e. only copy when the list is first modified during the dispatch?

**Comment by Hixie** [comment]
Looking at the benchmarks, it's not clear that this had the impact we were hoping for.

I didn't see any change on memory-related benchmarks I looked at, or on performance macrobenchmarks.

On the listener-specific microbenchmarks, on Android it made addListener2_iteration much (~10x) worse (maybe because it pushed the benchmark into needing more GC?), addListener3_iteration, addListener4_iteration, and addListener5_iteration a little worse (~15%), the notifyListenerX_iteration benchmarks got better (higher Xs got proportionally better improvement), and removeListenerX_iteration tests are mostly unaffected. On iOS the data is too noisy to discern an effect, though based on eyeballing the data, I suspect that if we were to do a rigorous analysis we'd find similar trends but with less amplitude.

Overall I think this was a wash. :-/ Probably worth keeping since hopefully adding listeners is rarer than notifying, though there's a chance this will have a negative effect on launch performance.

It might be that the increased constant cost of a linked list, and the loss of locality, offset much of the potential gain?

fyi @liyuqian

**Comment by rrousselGit** [comment]
Would it make sense to have a custom linked list implementation?

Another thing I've been thinking is, we could optimize 1-2 listeners by storing them out of the list:

```dart
class Notifier {
  VoidCallback firstListener;
  LinkedLisf<ListenerEntry> otherListeners;

  addListener(VoidCallback cb) {
    if (firstListener == null) {
      firstListener = cb;
    }
    else
       otherListeners.add(ListenerEntry(cb));
  } 
}
```

Thst could make very small listeners cases faster.

**Comment by letsar** [comment]
@Hixie I created another implementation based on a List and not a LinkedList. The benchmark result was pretty similar to what @rrousselGit came up, but it was always a bit behind its implementation so I didn't mention it.
How could I benchmark my implementation with the micro and macrobenchmark you mentioned before, so that I could see whether the list implementation brings performance improvements or not?

**Comment by Hixie** [comment]
They're the benchmarks should on https://flutter-dashboard.appspot.com/, which is just the tests listed in https://github.com/flutter/flutter/blob/master/dev/devicelab/manifest.yaml, which you can run locally, see the README in that directory.

**Comment by Kavantix** [comment]
@Hixie I have done some testing and I found that for some reason the iterator of `LinkedList` is very slow.
So that could explain why the `removeListener` benchmark became so much slower and also why the `notifyListeners` impact isn't as great as we hoped.

On my 7+ and iPad pro I ran a test in release mode where I do the notifyListeners pattern (list size is 2 and repeat is 1e6)

When using List.of it takes about 32 ms on the 7+ and 24ms on the ipad to do this:
```dart
  for (int j = 0; j < repeat; j++) {
    final entries = List.of(linked);
    for (final entry in entries) {
      if (entry.list == list) sum += entry.val;
    }
  }
```
(With a list size of 20 this becomes 173 ms on the 7+ and 105 ms on the ipad)

When I however use a custom copy of the list it takes only about 8ms on the 7+ and 5ms on the ipad:
```dart
  for (int j = 0; j < repeat; j++) {
    Entry entry = linked.first;
    final entries = List.generate(linked.length, (_) {
      final e = entry;
      entry = entry.next;
      return e;
    });
    for (final entry in entries) {
      if (entry.list == list) sum += entry.val;
    }
  }
```
(With a list size of 20 this becomes 72 ms on the 7+ and 40 ms on the ipad)

And for reference using a list instead of linked list:
```dart
  for (int j = 0; j < repeat; j++) {
    final copy = List.of(list);
    for (final i in copy) {
      if (list.contains(i)) sum += i;
    }
  }
```
This takes:
37 ms with size 2 and 436 ms with size 20 on the 7+
45 ms with size 2 and 294 ms with size 20 on the ipad

**Comment by liyuqian** [comment]
Traversal over a continuous range of elements is expected to be much slower for a `LinkedList` compared to a `List` for the following reasons.

- A `List` puts adjacent items in adjacent memory. When you access a list item, the CPU L1/L2 cache will fetch adjacent memory which likely covers the next few items. Hence visiting the next item will be a CPU L1/L2 cache hit which is much faster than accessing the main memory.
- A `LinkedList` puts adjacent item in random locations. Accessing an item is unlikely to fetch adjacent items into CPU L1/L2 cache. Hence visiting adjacent items will be slower due to cache miss, but it has a better asymptotic complexity when items can be inserted/removed at random locations. A `List` takes O(N) time to do that while a `LinkedList` takes O(1) time to do that.

Of course, one can implement a custom `HybridList` which links smaller continuous `List` together to utilize CPU L1/L2 cache for adjacent items, and limit worst case complexity by limiting the smaller `List` to have a constant size.

That may be an over-optimization unless we can prove that piece of code is actually a bottleneck in real applications. That said, I'm Ok with both the `List` and `LinkedList` approach since I haven't seen any related performance issues in real applications yet. Even the O(N^2) vs O(N) worst case complexity doesn't bother me too much as I expect N to be quite small (<10) in most cases. For example, the standard quick sort algorithm in C++ uses O(N^2) sorting algorithm instead of O(N log N) algorithm for small N.

In summary, if it's me, I'll just pick the one that's easiest to implement and read, until I see a performance issue in a real app.
